lib/grizzly/firmware_updates/otw_update_runner.ex

```diff
@@ -1,74 +1,655 @@
+    )
+
+    {:keep_state_and_data, [{:next_event, event_type, event_content}]}
+  end
+
+  #
+  # API CALLBACKS (call/cast/info)
+  #
+
+  # Ignore UART messages in other states
+  def handle_event(:info, {:circuits_uart, _port, data}, :init, data) do
+    :keep_state_and_data
+  end
+
+  # This is probably the ACK from the reboot to bootloader command.
+  def handle_event(:info, {:circuits_uart, _port, <<0x06>>}, :detect_bootloader_menu, _data) do
+    :keep_state_and_data
+  end
+
+  # We have the bootloader menu. Send "1" to start the upload.
+  def handle_event(
+        :info,
+        {:circuits_uart, _port, <<"\r\nGecko", _::binary>>},
+        :detect_bootloader_menu,
+        data
+      ) do
+    Process.sleep(10)
+    :ok = Circuits.UART.write(data.uart, "1\n")
+    :ok = Circuits.UART.drain(data.uart)
+    {:next_state, :uploading, data, []}
+  end
+
+  # Discard other UART data in detect_bootloader_menu state. It's probably SAPI commands,
+  # so send an ACK and another reboot command to keep things moving.
+  def handle_event(:info, {:circuits_uart, _port, _uart_data}, :detect_bootloader_menu, data) do
+    :ok = Circuits.UART.write(data.uart, <<0x06, 0x01, 0x03, 0x00, 0x27, 0xDB>>)
+    :ok = Circuits.UART.drain(data.uart)
+    :keep_state_and_data
+  end
+
+  # Ignore the "begin upload" message in uploading state
+  def handle_event(:info, {:circuits_uart, _port, "\r\nbegin upload\r\n"}, :uploading, _data) do
+    :keep_state_and_data
+  end
+
+  # Receive data from UART in uploading state
+  def handle_event(:info, {:circuits_uart, _port, uart_data}, :uploading, data) do
+    case Exmodem.receive_data(data.exmodem, uart_data) do
+      # When Exmodem tells us to send an ETB (end of transmission block), do so.
+      # However, instead of sending an ACK back, the Gecko bootloader will print
+      # "Serial upload complete" and then return to the bootloader prompt. At this
+      # point, we can stop the Exmodem process.
+      {:send, <<0x17>>} ->
+        Exmodem.stop(data.exmodem)
+        :ok = Circuits.UART.write(data.uart, <<0x17>>)
+        {:next_state, :upload_complete, %{data | exmodem: nil}, []}
+
+      {:send, to_send} ->
+        {sent, total} = Exmodem.progress(data.exmodem)
+
+        if rem(sent, div(total, 10)) == 0 do
+          Logger.info("[Grizzly] OTW update progress: #{sent}/#{total} packets sent")
+        end
+
+        :ok = Circuits.UART.write(data.uart, to_send)
+        :keep_state_and_data
+
+      :ignore ->
+        :keep_state_and_data
+
+      {:error, :canceled_by_receiver} ->
+        Logger.error("[Grizzly] Firmware upload canceled by receiver")
+        _ = module_reset(data)
+
+        {:next_state, :done, %{data | result: :update_rejected_by_zwave_module}}
+
+      {:error, :unexpected_data} ->
+        Logger.error(
+          "[Grizzly] Exmodem received unexpected data during upload: #{inspect(uart_data, limit: :infinity)}"
+        )
+
+        :keep_state_and_data
```

Comment:
Just curious: since we move to `:done` if `:canceled_by_receiver`, is there a reason we don't do that for `:unexpected_data`? What kind of "unexpected data" does this refer to?

Reply:
It should be exceedingly rare to hit this case. The most likely scenario is a bit getting flipped in transit over the UART resulting in an ACK/NACK from the receiver getting corrupted. If that happens, the receiver will retransmit the last ACK/NACK when it doesn't get the next packet from us in a timely manner.

Worst case scenario, hitting this case means the Z-Wave module got reset during transmission, in which case the unexpected data will either be SAPI commands or the bootloader menu. That's so unlikely that I didn't think it was really necessary to add code to handle the possibility -- eventually, we'll hit the state timeout and (most likely) retry from the top.